viewing_party/party.py

```diff
@@ -1,23 +1,201 @@
+        return {}
+
+    friends_watched = {}
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if subscription_filter:
+                if movie["host"] in user_data["subscriptions"]:
+                    friends_watched[movie["title"]] = movie
+            else:
+                friends_watched[movie["title"]] = movie
+    return friends_watched
+
+def get_unique_watched(user_data):
+    '''
+    determines movies user has watched,
+    friends have not
+
+    parameter: user_data(dictionary)
+    returns: list of movie dictionaries
+
+    '''
+    
+    unique_watched = []
+    friends_watched = get_friend_watched(user_data)
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_watched:
+            unique_watched.append(movie)
+
+    return unique_watched
+
+def get_friends_unique_watched(user_data):
+    '''
+    determines movies friends have watched,
+    user has not
+
+    parameter: user_data(dictionary)
+    returns: list of movie dictionaries
+    '''
+
+    friends_watched = get_friend_watched(user_data)
+    for movie in user_data["watched"]:
+        if movie["title"] in friends_watched:
+            friends_watched.pop(movie["title"])
+
+    return list(friends_watched.values())
+
+
 # ------------- WAVE 4 --------------------
-# -----------------------------------------
 
-# -----------------------------------------
+def get_available_recs(user_data):
+    '''
+    Determines list of recommended movies.
+    -User has not watched movies
+    -At least one friend has watched it
+    - The "host" of the movie is a service that is in the user's `"subscriptions"`
+
+    parameter: user_data(dictionary)
+
+    steps:
+    -find which movies friends have watched that user has not
+    -check movie “host” is under user's “subscriptions”
+    '''
+    friends_watched = get_friend_watched(user_data, subscription_filter=True)
+    for movie in user_data["watched"]:
+        if movie["title"] in friends_watched:
+            friends_watched.pop(movie["title"])
+    
+    return list(friends_watched.values())
+    
+
 # ------------- WAVE 5 --------------------
-# -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    '''
+    Determines list of recommended movies based on genre.
+    -User has not watched movies
+    -At least one friend has watched it
+    - Genre is user's most watched genre
+    parameter: user_data(dictionary)
+    return: list of movie
+    '''
+    recommendations = []
+
+    favorite_genre = get_most_watched_genre(user_data)
+
+    if not favorite_genre:
+        return []
+    
+    friends_watched = get_friend_watched(user_data)
+
+    for movie in user_data["watched"]:
+        if movie["title"] in friends_watched:
+            friends_watched.pop(movie["title"])
+
+    for movie in friends_watched.values():
+        if movie["genre"] == favorite_genre:
+            recommendations.append(movie)

```

Comment:
For loops like this are a great opportunity to use list comprehensions as you get more comfortable with Python! If we are filtering out specific values, list comprehensions are the way to go! That could look something like the following:

```suggestion
    recommendations = [movie for movie in friends_watched.values() if movie["genre"] == favorite_genre]

```